### DISCUSSION TITLE
Private positional and primary constructors

### DISCUSSION THREAD

**Discussion by jochenz** [dotnet/csharplang#7694] [Language Ideas]
Hi all,

records are a great way to represent "DDD" value types. For example, for a book ID, I could define:

```c#
public record BookId(int Value);
``` 

Very nice to have this concise syntax which gives me the readonly record.
A next step is that I wish to introduce validation for the value passed in. So I add a factory method:

```c#
public record BookId(int Value)
{
    public static BookId? TryCreate(int value)
    {
        // simplistic validation which avoids creating
        // invalid BookId instances (when this factory is used)
        return value <= 0 ? null : new BookId(value);
    }
}
```

Of course, the comment says it all: my validation is only used when the factory method is used. Anyone can still use the positional constructor and bypass validation. So, obviously, I want to make the constructor private. Which means, I need to make it an explicit constructor:

```c#
// BOOM: the whole thing explodes when I want a private constructor;
// this is basically how all my records look like, since I don't want anyone
// to create instances without input validation
public record BookId
{
    private BookId(int value)
    {
        Value = value;
    }

    public static BookId? TryCreate(int value)
    {
        return value <= 0 ? null : new BookId(value);
    }

    public int Value { get; init; }

    public void Deconstruct(out int value)
    {
        value = this.Value;
    }
}
```
Wouldn't it be nice to just protect against public usage of the constructor and to maintain the conciseness of the record? What about adding private to the positional (or primary) constructor?

**public record** BookId **_private_**(int Value)

This keeps all the super nice conciseness of the record type, but consumers can only create the type through the safe factory. I'd love to be able to do so :-)

```c#
public record BookId private(int Value)
{
    public static BookId? TryCreate(int value)
    {
        return value <= 0 ? null : new BookId(value);
    }
}
```

**Comment by HaloFour** [comment]
The team is interested in exploring primary constructor bodies or final initializers which would make it possible to support validation:  https://github.com/dotnet/csharplang/discussions/7667#discussioncomment-7536611
